### PR TITLE
Instead of sending a message to the index log when an indexing task finishes, stream results as they come in

### PR DESCRIPTION
**Pull request by ahoppen**
This also means that you can use the index log to view which tasks are currently being executed.

Since we only have a single log stream we can write to, I decided to prefix every line in the index log with two colored emojis that an easy visual association of every log line to the task that generated them.

This is how an index log now looks like in VS Code

```
[Info  - 9:49:31 AM] ⬜️🟥 [1/1] Compiling plugin Format Source Code
[Info  - 9:49:31 AM] ⬜️🟥 Finished in 0.51599975 seconds
[Info  - 9:49:31 AM] 🟥🟥 Preparing GenerateManual for dummy
/Library/Developer/Toolchains/swift-latest.xctoolchain/usr/bin/swift build --package-path /Users/alex/src/sourcekit-lsp --scratch-path /Users/alex/src/sourcekit-lsp/.index-build --disable-index-store --target GenerateManual
[Info  - 9:49:32 AM] 🟧🟩 Indexing /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-format/Plugins/FormatPlugin/plugin.swift
/Library/Developer/Toolchains/swift-latest.xctoolchain/usr/bin/swiftc -swift-version 5 -I /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/pm/ManifestAPI -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.4.sdk -package-description-version 5.6.0 /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-format/Plugins/FormatPlugin/plugin.swift -index-file -index-file-path /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-format/Plugins/FormatPlugin/plugin.swift -disable-batch-mode -index-unit-output-path /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-format/Plugins/FormatPlugin/plugin.swift.o
[Info  - 9:49:32 AM] 🟧🟩 /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-format/Plugins/FormatPlugin/plugin.swift:1:8: error: no such module 'PackagePlugin'
[Info  - 9:49:32 AM] 🟧🟩  1 | import PackagePlugin
[Info  - 9:49:32 AM] 🟧🟩    |        `- error: no such module 'PackagePlugin'
[Info  - 9:49:32 AM] 🟧🟩  2 | import Foundation
[Info  - 9:49:32 AM] 🟧🟩  3 | 
[Info  - 9:49:32 AM] 🟧🟩 Finished in 0.205038459 seconds
[Info  - 9:49:32 AM] 🟥🟥 [1/1] Compiling plugin GenerateManual
[Info  - 9:49:32 AM] 🟥🟥 Finished in 0.522686625 seconds
[Info  - 9:49:32 AM] 🟦🟨 Preparing _CryptoExtras for dummy
/Library/Developer/Toolchains/swift-latest.xctoolchain/usr/bin/swift build --package-path /Users/alex/src/sourcekit-lsp --scratch-path /Users/alex/src/sourcekit-lsp/.index-build --disable-index-store --target _CryptoExtras
[Info  - 9:49:32 AM] 🟩⬜️ Indexing /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-argument-parser/Plugins/GenerateManual/GenerateManualPlugin.swift
/Library/Developer/Toolchains/swift-latest.xctoolchain/usr/bin/swiftc -swift-version 5 -I /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/pm/ManifestAPI -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.4.sdk -package-description-version 5.6.0 /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-argument-parser/Plugins/GenerateManual/GenerateManualPlugin.swift /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-argument-parser/Plugins/GenerateManual/GenerateManualPluginError.swift /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-argument-parser/Plugins/GenerateManual/PackagePlugin+Helpers.swift -index-file -index-file-path /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-argument-parser/Plugins/GenerateManual/GenerateManualPlugin.swift -disable-batch-mode -index-unit-output-path /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-argument-parser/Plugins/GenerateManual/GenerateManualPlugin.swift.o
[Info  - 9:49:32 AM] 🟧⬛️ Indexing /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-argument-parser/Plugins/GenerateManual/GenerateManualPluginError.swift
/Library/Developer/Toolchains/swift-latest.xctoolchain/usr/bin/swiftc -swift-version 5 -I /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/pm/ManifestAPI -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.4.sdk -package-description-version 5.6.0 /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-argument-parser/Plugins/GenerateManual/GenerateManualPlugin.swift /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-argument-parser/Plugins/GenerateManual/GenerateManualPluginError.swift /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-argument-parser/Plugins/GenerateManual/PackagePlugin+Helpers.swift -index-file -index-file-path /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-argument-parser/Plugins/GenerateManual/GenerateManualPluginError.swift -disable-batch-mode -index-unit-output-path /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-argument-parser/Plugins/GenerateManual/GenerateManualPluginError.swift.o
[Info  - 9:49:32 AM] 🟩🟪 Indexing /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-argument-parser/Plugins/GenerateManual/PackagePlugin+Helpers.swift
/Library/Developer/Toolchains/swift-latest.xctoolchain/usr/bin/swiftc -swift-version 5 -I /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/pm/ManifestAPI -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.4.sdk -package-description-version 5.6.0 /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-argument-parser/Plugins/GenerateManual/GenerateManualPlugin.swift /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-argument-parser/Plugins/GenerateManual/GenerateManualPluginError.swift /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-argument-parser/Plugins/GenerateManual/PackagePlugin+Helpers.swift -index-file -index-file-path /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-argument-parser/Plugins/GenerateManual/PackagePlugin+Helpers.swift -disable-batch-mode -index-unit-output-path /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-argument-parser/Plugins/GenerateManual/PackagePlugin+Helpers.swift.o
[Info  - 9:49:32 AM] 🟩🟪 /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-argument-parser/Plugins/GenerateManual/GenerateManualPlugin.swift:12:8: error: no such module 'PackagePlugin'
[Info  - 9:49:32 AM] 🟩🟪 10 | //===----------------------------------------------------------------------===//
[Info  - 9:49:32 AM] 🟩🟪 11 | 
[Info  - 9:49:32 AM] 🟩🟪 12 | import PackagePlugin
[Info  - 9:49:32 AM] 🟩🟪    |        `- error: no such module 'PackagePlugin'
[Info  - 9:49:32 AM] 🟩🟪 13 | 
[Info  - 9:49:32 AM] 🟩🟪 14 | @main
[Info  - 9:49:32 AM] 🟩⬜️ /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-argument-parser/Plugins/GenerateManual/GenerateManualPlugin.swift:12:8: error: no such module 'PackagePlugin'
[Info  - 9:49:32 AM] 🟩⬜️ 10 | //===----------------------------------------------------------------------===//
[Info  - 9:49:32 AM] 🟩⬜️ 11 | 
[Info  - 9:49:32 AM] 🟩⬜️ 12 | import PackagePlugin
[Info  - 9:49:32 AM] 🟩⬜️    |        `- error: no such module 'PackagePlugin'
[Info  - 9:49:32 AM] 🟩⬜️ 13 | 
[Info  - 9:49:32 AM] 🟩⬜️ 14 | @main
[Info  - 9:49:32 AM] 🟧⬛️ /Users/alex/src/sourcekit-lsp/.index-build/checkouts/swift-argument-parser/Plugins/GenerateManual/GenerateManualPlugin.swift:12:8: error: no such module 'PackagePlugin'
[Info  - 9:49:32 AM] 🟧⬛️ 10 | //===----------------------------------------------------------------------===//
[Info  - 9:49:32 AM] 🟧⬛️ 11 | 
[Info  - 9:49:32 AM] 🟧⬛️ 12 | import PackagePlugin
[Info  - 9:49:32 AM] 🟧⬛️    |        `- error: no such module 'PackagePlugin'
[Info  - 9:49:32 AM] 🟧⬛️ 13 | 
[Info  - 9:49:32 AM] 🟧⬛️ 14 | @main
[Info  - 9:49:32 AM] 🟩⬜️ Finished in 0.159845792 seconds
[Info  - 9:49:32 AM] 🟧⬛️ Finished in 0.159851625 seconds
[Info  - 9:49:32 AM] 🟩🟪 Finished in 0.159853 seconds
```

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

File: Sources/SourceKitLSP/SourceKitLSPServer.swift [hunk 1]
```diff
@@ -931,8 +947,8 @@ extension SourceKitLSPServer {
       options: options,
       indexOptions: self.options.indexOptions,
       indexTaskScheduler: indexTaskScheduler,
-      indexProcessDidProduceResult: { [weak self] in
-        self?.indexTaskDidProduceResult($0)
+      logMessageToIndexLog: { taskID, message in
+        self.logMessageToIndexLog(taskID: taskID, message: message)
```
Comment (hamishknight) [inline Sources/SourceKitLSP/SourceKitLSPServer.swift:951]: No `weak self`?

File: Sources/SourceKitLSP/SourceKitLSPServer.swift [hunk 2]
```diff
@@ -1012,8 +1028,8 @@ extension SourceKitLSPServer {
           index: nil,
           indexDelegate: nil,
           indexTaskScheduler: self.indexTaskScheduler,
-          indexProcessDidProduceResult: { [weak self] in
-            self?.indexTaskDidProduceResult($0)
+          logMessageToIndexLog: { taskID, message in
+            self.logMessageToIndexLog(taskID: taskID, message: message)
```
Comment (hamishknight) [inline Sources/SourceKitLSP/SourceKitLSPServer.swift:1032]: Same comment

File: Sources/SourceKitLSP/SourceKitLSPServer.swift (hunk 1 above)
Comment (hamishknight) [reply Sources/SourceKitLSP/SourceKitLSPServer.swift:951]: Also, should have made this comment sooner, but should we consider using a delegate or something to cut down on the amount of closures we have to pass around?

File: Sources/SKSupport/PipeAsStringHandler.swift [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  private let handler: @Sendable (String) -> Void
+
+  public init(handler: @escaping @Sendable (String) -> Void) {
+    self.handler = handler
+  }
+
+  private func handleDataFromPipeImpl(_ newData: Data) {
+    self.buffer += newData
+    while let newlineIndex = self.buffer.firstIndex(of: UInt8(ascii: "\n")) {
+      // Output a separate log message for every line in clangd's stderr.
```
Comment (hamishknight) [inline Sources/SKSupport/PipeAsStringHandler.swift:33]: I think this comment probably needs updating

File: Sources/SKSupport/PipeAsStringHandler.swift [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  /// The closure that actually handles
+  private let handler: @Sendable (String) -> Void
+
+  public init(handler: @escaping @Sendable (String) -> Void) {
+    self.handler = handler
+  }
+
+  private func handleDataFromPipeImpl(_ newData: Data) {
+    self.buffer += newData
+    while let newlineIndex = self.buffer.firstIndex(of: UInt8(ascii: "\n")) {
```
Comment (hamishknight) [inline Sources/SKSupport/PipeAsStringHandler.swift:32]: Scanning the whole buffer every time doesn't seem ideal, but I guess lines aren't that long in practice?

**Review by hamishknight — APPROVED** [review]

**Review by hamishknight — APPROVED** [review]

File: Sources/SKSupport/PipeAsStringHandler.swift (hunk 4 above)
Comment (ahoppen) [reply Sources/SKSupport/PipeAsStringHandler.swift:32]: Yes, the expectation is that lines aren’t long and I can’t think of common scenarios where you get less than a full line from a pipe.

File: Sources/SourceKitLSP/SourceKitLSPServer.swift [hunk 5]
```diff
@@ -931,8 +947,8 @@ extension SourceKitLSPServer {
       options: options,
       indexOptions: self.options.indexOptions,
       indexTaskScheduler: indexTaskScheduler,
-      indexProcessDidProduceResult: { [weak self] in
-        self?.indexTaskDidProduceResult($0)
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SourceKitLSP/SourceKitLSPServer.swift:936]: Oh, yes. Thanks for catching! Could also remove `IndexProcessResult` now.

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Comment by ahoppen** [comment]
@swift-ci Please test macOS

**Pull request merged**